### DISCUSSION TITLE
How would you explain the benefit of the Monadic interface when implemented for Reactive structures to someone unfamiliar with Monads?

### DISCUSSION THREAD

**Discussion by ken-okabe** [fluture-js/Fluture#568] [Q&A]
I just wonder what is the advantage of having a reactive structure as monads rather than functors.
For list or option, it's relatively well-known how good it is to have monads rather than functors.
Can anyone explain this with a simple example? Thanks.

**Comment by Avaq** [comment]
I don't understand your question. I'm going to attempt to pick it apart, in the hopes that I'll answer your question in the process of doing so.

### 1. "Monads rather than Functors"

You've said the phrase "monads rather than functors" a few times in your question, but *Monads are Functors*:

![](https://raw.githubusercontent.com/fantasyland/fantasy-land/master/figures/dependencies.png)

As seen from the diagram above, Monads are a combination of Functor, Apply/Applicative, and Apply/Chain. So rather than asking "Why choose a Monad *rather than* a Functor?", you should ask "Can and should my Functor be a Monad as well?". I'll answer the latter question below.

### 2. Should my Functor be a Monad?

A Monad is the generalization of the "updating a value in sequence" (whereas a functor is just "updating a value"). When I say "in sequence", it means that the value was generated with an effect, and the updated value is *also* created with an effect of the same nature, and these effects *take place in sequence*.

A Functor doesn't carry the capability for composing two effectful computations into one. Instead, a Functor takes one effectful computation, and one pure computation, and composes those.

You can see how composition of two effectful computations is relevant for abstractions where the effect is asynchronousity (like Fluture): One value is created asynchronously, and then based on that value, a new value is created asynchronously as well, and we want those two asynchronous computations to be composed such that it becomes a single asynchronous computation.

### 3. Explain this with a simple example

It's not clear from your question exactly what you want explained. But I'll give an example for the above answer.

```js
// Here we create a basic Functor (with no effect)
const MyFunctor = value => ({
  run: () => value,
  map: f => MyFunctor(f(value)),
});

// We can now compose an effectful value with a pure computation
// to create a new effectful value:
const pureComputation = x => x + 1;
const effectful = MyFunctor(41);
const newEffectful = effectful.map(pureComputation);

// We end up with a single new effectful value:
effectful.run() // 41
newEffectful.run() // 42

// But if we try to compose with an effectful computation:
const effectfulComputation = x => MyFunctor(x + 1);
const newEffectful2 = effectful.map(effectfulComputation);

// We end up with a nested effectful value, meaning we didn't really "compose"
// the computation with the effect, because what came out was not of the same
// type as what we put in. We have to run it twice:
newEffectful2.run().run() // 42

// Now let's make composition with the effectful computation possible.
// We create a Monad implementation that knows how to run the effect on the
// inside as part of its new effect:
const MyMonad = value => ({
  ...MyFunctor(value),
  chain: f => MyFunctor(f(value).run())
});

// Now we can compose the effectful value with an effectful computation:
const effectful3 = MyMonad(41);
const effectfulComputation3 = x => MyMonad(x + 1);
const newEffectful3 = effectful3.chain(effectfulComputation3);

// And we get a new effectful value that has the same type:
newEffectful3.run() // 42
```

### 4. What is the simplest usage of Fluture monad?

The title of your question suggests that you're looking for something else entirely. For a very simple example of how one might use the Monadic interface as implemented in Fluture, you can refer to the [documentation on `chain`](https://github.com/fluture-js/Fluture#chain).

----

I hope this helps! If it doesn't answer your question, please elaborate a little bit.

**Reply by ken-okabe** [reply]
@Avaq Thank you for your detailed answer on this topic. I apologize for taking up your time with my question with insufficient information.
Probably, I think I understand your description thoroughly.

I am looking for useful ways to use monads in practical applications of FRP.

In functional programming, much is written about List-Monad or Option(Maybe)-Monad and Parser-Monad are very impressive.

------

For instance, [Array.flatMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap) is a List-Monad in JavaScript.

A simple yet useful example of using `Array.flatMap` is the ability to manipulate arrays without the need for complex native Array methods such as `filter` or `find`.

The code below illustrates how to remove a specific word from text by utilizing `Array.flatMap` alone:
```ts
type remove = (word: string) => (list: string[]) => string[]; 
const remove: remove =
    word => list =>
        list.flatMap(
            a =>
                a === word
                    ? []  // remove the target word
                    : [a] // preserve the exisiting word
        );

const list1 = ["apple", "orange", "banana"];
const list2 = remove("orange")(list1);

console.log(list1); // [ 'apple', 'orange', 'banana' ]
console.log(list2); // [ 'apple', 'banana' ]
```

It is possible because `Array.flatMap` is a monad, whereas it is not possible with `Array.map` which is only a functor.
So this is a good example that anyone can see how powerful the List-Monad: `Array.flatMap` can be.

----

Now, for Functional Reactive Programming(FRP).

I implemented my own Reactive Monad.
https://gist.github.com/stken2050/2fe649f9a8499ef305c4db026aa01c51

An example code would be:
```ts
  import { R } from "./reactive_monad.js";
  import type { Reactive } from "./reactive_monad.js";

  type log = <A>(a: A) => A;
  const log: log = a => {
    console.log(a);
    return a;
  };
  //--------------------------
  type f = (a: number) => number;
  const f: f = a => a * 2;

  type monadF = (a: number) => Reactive<number>;
  const monadF: monadF = a => R(a * 2);
  //--------------------------
  R("reactiveVal will be initialized: 1").map(log);
  const reactiveVal = R(1); // reactiveVal is initialized: 1
  //--------------------------
  const result = reactiveVal.map(f); // Functor
  const resultM = reactiveVal.flatMap(monadF); // Monad
  //--------------------------
  result.map(log);  // 2 then later 10
  resultM.map(log); // 2 then later 10
  //--------------------------
  R("reactiveVal will be updated: 1 -> 5").map(log);
  reactiveVal.next(5);  // reactiveVal is updated: 1 -> 5
```

console output:

```sh
reactiveVal will be initialized: 1
2
2
reactiveVal will be updated: 1 -> 5
10
10
```

This works as expected; however, the problem is so far, I cannot find any good example to show off how powerful the ` reactiveVal.flatMap(monadF)` can be, like `array.flatMap(monadF)`. In other words, I can not find a good case which is possible because `Reactive.flatMap` is a monad, whereas it is not possible with `Reactive.map` which is only a functor.

In fact, I understand pretty well in terms of algebraic structure:
>end up with a nested effectful value, meaning we didn't really "compose"

However, this concept is too abstract and may not be accessible to beginners or those looking for practical applications.

Do you have any idea?

**Reply by Avaq** [reply]
Based on your elaboration, I think your question could be rephrased as: *How would you explain the benefit of the Monadic interface when implemented for Reactive structures to someone unfamiliar with Monads?* Is that right? If so, could you update the question title as well? I'll post a new answer.

**Reply by ken-okabe** [reply]
> Based on your elaboration, I think your question could be rephrased as: _How would you explain the benefit of the Monadic interface when implemented for Reactive structures to someone unfamiliar with Monads?_ Is that right? If so, could you update the question title as well? I'll post a new answer.

I am familiar with Monads and am able to implement a Monadic interface for a Reactive structure on my own, however, I am still uncertain of the benefits. Nevertheless, I have updated the question title as the answer should remain the same. Thank you.

**Comment by Avaq** [comment]
## What are monads?

The Monadic interface is one of several "algebraic" interfaces inspired by "category theory".

Category theory is simply said *a theory of theories*: It is used to categorize other mathematical theories, and itself. Just as category theory is a theory of theories, the programming interfaces inspired by it are *abstractions of abstraction*.

Each of these interfaces categorize a common abstraction in programming. For example, the [Monoid interface](https://github.com/fantasyland/fantasy-land#monoid) is the generalization of binary operators that combine two values in a specific way. The plus operator is a Monoid for number, the `.concat` method is a Monoid for Array, and there are many more things that can be categorized as Monoids.

With our Monoid example in mind, let's examine what the Monad interface is a categorization of. It's a little bit more abstract: It's basically the generalization of any operation that depends on the result of a previous, similar operation. So any time we go: "Do one thing, and depending on how that turns out, do another thing", we're doing something Monadic.

We can now make the mental jump to Promises, which are a typical example of a Monadic abstraction. Their whole interface (`.then`) is modelled to let a user do one thing, and _then_ do something with the result of that. But there are many other things that could be categorized as Monadic, even for example imperative programming languages _themselves_. The Monadic interface is extremely broadly applicable!

## Why make monads?

So why would we go through the trouble of making our abstractions that _could_ be Monadic, actually conform to a common Monadic interface such as [the one specified in Fantasy Land](https://github.com/fantasyland/fantasy-land#monad)?

The primary advantages from my perspective are:

- Reduced mental overhead and cost of learning an abstraction, and
- interoperability with other code.

When all Monadic abstractions conform to a common monadic interface, then learning them becomes real easy! You _just_ need to learn the monadic interface once, and suddenly you can use anything and everything that has "do one thing, and then use its result" at its core: Asynchronous things, optional things, streamed data, parsing, state! The list goes on. Furthermore, because this interface is so strictly defined, it doesn't require much thought to work with it. We know exactly what to expect from a Monad.

Finally, it's now trivial to write very abstract code that doesn't _just_ work on Promises, or _just_ work on Streams, but works on all Monadic abstractions! For example, I can write the following function:

```js
const andThen = (firstMonad, secondMonad) => firstMonad.chain(() => secondMonad);
```

And instantly I have a function that sequences two operations neatly after one another. It works with Promises, in which case the first async thing happens, and when it's done, the second one happens. But it works also with Optional values, in which case if the first thing doesn't happen, the second one also won't.

This ability to abstract over so many abstractions at once goes much further too. We can create (and have created) general utilities for combining multiple Monads into one, in a fully automated way. Where I can just do `OptionalPromise = OptionT(Promise)` or something to create Promises that optionally execute. The possibilities are endless!

I hope this convinces you to try and conform your abstractions to the Monadic interface, and/or some of the many other algebraic interfaces.
